In [208]:
import pandas as pd
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

In [105]:
# all variables
fb_train = pd.read_csv("./facebook_train.csv")
fb_train = pd.concat([fb_train, pd.get_dummies(fb_train["type"]), pd.get_dummies(fb_train["category"]), pd.get_dummies(fb_train["weekday"]), pd.get_dummies(fb_train["month"]), pd.get_dummies(fb_train["hour"])], axis=1)
fb_train = fb_train.drop(["type", "category", "month", "weekday", "hour"], axis=1)
fb_train.columns = ["followers", "paid", "comments", "likes", "shares", "Link", "Photo", "Status", "Video", "Category 1", "Category 2", "Category 3", "Mon","Tue","Wed","Thu","Fri","Sat","Sun", "Jan","Feb","Mar","May","Apr","Jun", "Jul", "Aug","Sep","Oct","Nov", "Dez", "1am", "2am", "3am", "4am", "5am", "6am", "7am", "8am", "9am", "12pm", "3pm", "4pm", "5pm", "6pm", "7pm", "8pm", "9pm","10pm", "11pm","12am"]

In [36]:
#no hours and above
fb_train = pd.read_csv("./facebook_train.csv")
fb_train = pd.concat([fb_train, pd.get_dummies(fb_train["type"]), pd.get_dummies(fb_train["category"]), pd.get_dummies(fb_train["weekday"]), pd.get_dummies(fb_train["month"])], axis=1)
fb_train = fb_train.drop(["type", "category", "month", "weekday", "hour"], axis=1)
fb_train.columns = ["followers", "paid", "comments", "likes", "shares", "Link", "Photo", "Status", "Video", "Category 1", "Category 2", "Category 3", "Mon","Tue","Wed","Thu","Fri","Sat","Sun", "Jan","Feb","Mar","May","Apr","Jun", "Jul", "Aug","Sep","Oct","Nov", "Dez"]

In [41]:
#no month and above
fb_train = pd.read_csv("./facebook_train.csv")
fb_train = pd.concat([fb_train, pd.get_dummies(fb_train["type"]), pd.get_dummies(fb_train["category"]), pd.get_dummies(fb_train["weekday"])], axis=1)
fb_train = fb_train.drop(["type", "category", "month", "weekday", "hour"], axis=1)
fb_train.columns = ["followers", "paid", "comments", "likes", "shares", "Link", "Photo", "Status", "Video", "Category 1", "Category 2", "Category 3", "Mon","Tue","Wed","Thu","Fri","Sat","Sun"]

In [185]:
#no day of the week and above
fb_train = pd.read_csv("./facebook_train.csv")
fb_train = pd.concat([fb_train, pd.get_dummies(fb_train["type"]), pd.get_dummies(fb_train["category"])], axis=1)
fb_train = fb_train.drop(["type", "category", "month", "weekday", "hour"], axis=1)
fb_train.columns = ["followers", "paid", "comments", "likes", "shares", "Link", "Photo", "Status", "Video", "Category 1", "Category 2", "Category 3"]
print(fb_train)

     followers  paid  comments  likes  shares  Link  Photo  Status  Video  \
0       124050     1         3     77      16     0      1       0      0   
1       132556     0        11    313      50     0      1       0      0   
2        92786     1         2    145      39     0      1       0      0   
3       135632     0         2    299      49     0      1       0      0   
4       139540     1         7    229      21     1      0       0      0   
..         ...   ...       ...    ...     ...   ...    ...     ...    ...   
392     138114     0        12    498      87     0      0       1      0   
393     127937     0         4     85      20     0      0       1      0   
394     135656     0         7    397     143     0      0       0      1   
395     124663     1         1     87      18     0      1       0      0   
396      93096     1         6    296      32     0      1       0      0   

     Category 1  Category 2  Category 3  
0             0           0      

In [100]:
#only paid
fb_train = pd.read_csv("./facebook_train.csv")
fb_train = pd.concat([fb_train, pd.get_dummies(fb_train["type"])], axis=1)
fb_train = fb_train.drop(["total_followers", "type", "category", "month", "weekday", "hour","paid"], axis=1)
fb_train.columns = ["comments", "likes", "shares","Link", "Photo", "Status", "Video"]

In [186]:
#splitting model
fb_model = fb_train.drop(["comments", "shares"], axis=1)
fb_target = fb_model.likes
x_train, x_test, y_train, y_test = train_test_split(fb_model, fb_target, test_size=0.2, random_state=0)
x_train = x_train.drop(["likes"], axis=1)
y_train

64       89
55       18
316     430
102    2306
261     122
       ... 
323     618
192      67
117    1479
47      421
172     423
Name: likes, Length: 317, dtype: int64

In [187]:
#running regression and cross validation
reg = linear_model.LinearRegression()
scores = cross_val_score(reg, x_train, y_train, cv=10)
print(scores)
reg.fit(x_train, y_train)

[ 0.07946035  0.18919212 -0.60292412 -0.37491044  0.04876056 -0.80796904
 -0.24722036  0.1040883  -0.01501424 -0.10691361]


LinearRegression()

In [148]:
#displaying coefficients for linear model
results_df = pd.DataFrame({"var": fb_model.drop(["likes"], axis=1).columns, "coef": reg.coef_})

results_df.round(2).sort_values('coef',ascending=False)

var    coef
42         7pm  647.67
16         Jan  581.07
17         Feb  415.74
18         Mar  216.70
31         4am  209.01
34         7am  175.54
19         May  158.09
1         paid  102.86
15         Sun   98.36
14         Sat   63.47
7   Category 2   61.10
5        Video   46.79
3        Photo   44.56
20         Apr   35.91
47        12am   31.85
13         Fri   25.22
35         8am   15.30
9          Mon   13.31
8   Category 3    9.76
0    followers    0.02
33         6am    0.00
40         5pm   -0.23
29         2am  -15.44
41         6pm  -19.04
2         Link  -35.59
30         3am  -37.25
22         Jul  -38.59
32         5am  -44.15
43         8pm  -47.63
39         4pm  -50.55
4       Status  -55.75
11         Wed  -58.97
10         Tue  -59.10
28         1am  -62.79
6   Category 1  -70.85
37        12pm  -72.18
46        11pm  -73.84
12         Thu  -82.28
36         9am -124.26
44         9pm -126.36
25         Oct -146.35
45        10pm -171.78
21         Jun -182.77
24         Sep -214.09
23         Aug -223.37
38         3pm -233.88
27         Dez -290.02
26         Nov -312.32

In [149]:
#showing results
print(mean_squared_error(y_test, reg.predict(x_test.drop(["likes"], axis=1))))
print(mean_absolute_error(y_test, reg.predict(x_test.drop(["likes"], axis=1))))
print(r2_score(y_test, reg.predict(x_test.drop(["likes"], axis=1))))

131740.65351615925
212.83526390027322
-0.24350268549993337


In [190]:
fb_test = pd.read_csv('./facebook_test.csv')
# all variables
fb_test = pd.concat([fb_test, pd.get_dummies(fb_test["type"]), pd.get_dummies(fb_test["category"])], axis=1)
fb_test = fb_test.drop(["type", "category", "month", "weekday", "hour"], axis=1)
fb_test.columns = ["followers", "paid", "comments", "likes", "shares", "Link", "Photo", "Status", "Video", "Category 1", "Category 2", "Category 3"]
print(fb_test)
r2_score(fb_test["likes"], reg.predict(fb_test.drop(["likes", "comments", "shares"], axis=1)))

ValueError: Length mismatch: Expected axis has 11 elements, new values have 12 elements

# Interpretation

The linear model is performing very poorly as suggested by the cross-validation test and the negative scores – a negative score indicates that simply taking the mean would perform better than the model itself. This is very likely since there shouldn't be a linear relationship between the number of likes for a given post and all the other variables. Moreover, some of the variables are not complete (such as time of the day).

To check if this occurs due to model complexity I ran the regression with multiple number of reduced parameters. The more parameters I withhold, the better the prediction. However, this assumes a linear relationship which I am then just optimizing for. Instead it would have been better to find a non-polynomial or n-degree polynomial model that better represents the underlying relationship and bias.

When running the regression with all parameters, the following seems to influence the number of likes the most:

- Either early morning or evening (but not too late)
- Beginning of the year
- End of the week
- Paid
- Category 2
- Video or Photo

The opposite of that seems to reduce likes the most (aka late in the evening, mid to end of the year).

In [115]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier 

In [164]:
#loading and inspecting data
casual_data = pd.read_csv("./casualty_train.csv")
casual_data = casual_data[(casual_data['age'] != 'Unknown')]
casual_data = casual_data[(casual_data['pedestrian_location'] != 'Unknown')]
casual_data["severe"].describe()
#print(casual_data.pedestrian_location.unique())
#print(casual_data.pedestrian_movement.unique())

count     214126
unique         2
top        False
freq      191159
Name: severe, dtype: object

In [152]:
#all variables
casual_data_dummy = pd.concat([casual_data, pd.get_dummies(casual_data["casualty_class"]), pd.get_dummies(casual_data["gender"]), pd.get_dummies(casual_data["travel"]), pd.get_dummies(casual_data["pedestrian_location"]), pd.get_dummies(casual_data["pedestrian_movement"]), pd.get_dummies(casual_data["year"])], axis=1).drop(["casualty_class", "gender", "pedestrian_location", "pedestrian_movement", "travel","year"],axis=1)

In [153]:
#no pedestrian movement
casual_data_dummy = pd.concat([casual_data, pd.get_dummies(casual_data["casualty_class"]), pd.get_dummies(casual_data["gender"]), pd.get_dummies(casual_data["travel"]), pd.get_dummies(casual_data["pedestrian_location"]), pd.get_dummies(casual_data["year"])], axis=1).drop(["casualty_class", "gender", "pedestrian_location", "pedestrian_movement", "travel","year"],axis=1)

In [154]:
#no pedestrian location and above
casual_data_dummy = pd.concat([casual_data, pd.get_dummies(casual_data["casualty_class"]), pd.get_dummies(casual_data["gender"]), pd.get_dummies(casual_data["travel"]), pd.get_dummies(casual_data["year"])], axis=1).drop(["casualty_class", "gender", "pedestrian_location", "pedestrian_movement", "travel","year"],axis=1)

In [155]:
#no year and above
casual_data_dummy = pd.concat([casual_data, pd.get_dummies(casual_data["casualty_class"]), pd.get_dummies(casual_data["gender"]), pd.get_dummies(casual_data["travel"])], axis=1).drop(["casualty_class", "gender", "pedestrian_location", "pedestrian_movement", "travel","year"],axis=1)

In [156]:
#data splitting
x_train, x_test, y_train, y_test = train_test_split(casual_data_dummy.drop(["severe"], axis=1), casual_data_dummy.severe, test_size=0.2, random_state=0)
x_train.columns

Index(['age', 'driver', 'passenger', 'pedestrian', 'female', 'male', 'bicycle',
       'bus', 'car', 'motorbike', 'other', 'pedestrian', 'taxi'],
      dtype='object')

In [160]:
#running classification
rf = RandomForestClassifier()
rf.fit(x_train, y_train)
rf.predict(x_test)
#print(rf.score(x_test, y_test))

array([False, False, False, ..., False, False, False])

In [158]:
#showing variable importance
feature_importances = pd.DataFrame(rf.feature_importances_, index = x_train.columns, columns=['importance'])
feature_importances

importance
age           0.444847
driver        0.025597
passenger     0.025116
pedestrian    0.129880
female        0.011688
male          0.012966
bicycle       0.101521
bus           0.007733
car           0.155882
motorbike     0.048790
other         0.005406
pedestrian    0.022619
taxi          0.007955

In [207]:
c_test = pd.read_csv('./casualty_test.csv')
print(c_test.severe.describe())
c_test = c_test[(c_test['age'] != 'Unknown')]
c_test = c_test[(c_test['pedestrian_location'] != 'Unknown')]
c_test_dummy = pd.concat([c_test, pd.get_dummies(c_test["casualty_class"]), pd.get_dummies(c_test["gender"]), pd.get_dummies(c_test["travel"])], axis=1).drop(["casualty_class", "gender", "pedestrian_location", "pedestrian_movement", "travel","year"],axis=1)
print(rf.score(c_test_dummy.drop(["severe"], axis=1), c_test_dummy["severe"]))

count     58208
unique        2
top       False
freq      51935
Name: severe, dtype: object
0.8915072371975187


In [210]:
conf_matrix = confusion_matrix(c_test_dummy["severe"], rf.predict(c_test_dummy.drop(["severe"], axis=1)))
conf_matrix

array([[47849,    42],
       [ 5782,     8]])

# Interpretation

The Random Forest Classifier is performing relatively well (89% accuracy). Since a Random Forest basically acts like cross-validation, I did not perform cross-validation for this classification. To improve the classification, I removed several variable classes such as the pedestrian movement, the pedestrian location, and the year.

As seen in the table above, the age of the person as well as whether he was the pedestrian or traveling in a car or bicycle mattered most. This is very intuitive as the older the person is, the more severe the casualty should be. Additionally, if one is the pedestrian and is hit by a car or bicycle, it will be much more severe (and also likely).